RL.py

```diff
@@ -0,0 +1,274 @@
+                v_batch = CACTO.critic_model(state_batch_norm, training=True)                      # Compute batch of Values associated to the sampled batch ofstates
+                v_next_batch = CACTO.target_critic(state_next_batch_norm, training=True)               # Compute batch of Values from target critic associated to sampled batch of next states
+                vref_batch = reward_batch + (1-d_batch)*(v_next_batch)                           # Compute the targets for the TD error         
+            else:
+                v_batch = CACTO.critic_model(state_batch, training=True) 
+                v_next_batch = CACTO.target_critic(state_next_batch, training=True)                     
+                vref_batch = reward_batch + (1-d_batch)*(v_next_batch)                          
+
+            td_errors = tf.math.abs(tf.math.subtract(vref_batch,v_batch))                           
+            new_priorities = td_errors.numpy() + self.conf.prioritized_replay_eps                      # Proportional prioritization where p_i = |TD_error_i| + self.conf.prioritized_replay_eps 
+            prioritized_buffer.update_priorities(batch_idxes, new_priorities)   
+
+        # Update NNs
+        self.update(episode, state_batch_norm, state_batch, state_batch_norm, state_batch, reward_batch, state_next_batch, state_next_batch_norm, d_batch, action_batch, weights_batch=weights_batch)
+
+    def RL_Solve(self, state_prev, ep, rand_time, env, tau_TO, prioritized_buffer):
+        ''' Solve RL problem '''
+        DONE = 0                              # Flag indicating if the episode has terminated
+        ep_return = 0                         # Initialize the return
+        step_counter = 0                      # Initialize the counter of episode steps
+        ep_arr = []
+
+        # START RL EPISODE
+        while True:
+            # Get current and next TO actions
+            action, next_TO_action = self.get_TO_actions(step_counter, action_TO=tau_TO)           
+
+            # Simulate actions and retrieve next state and compute reward
+            state_next, rwrd = env.step(rand_time, state_prev, action)
+            
+            # Store performed action and next state and reward
+            CACTO.control_arr = np.vstack([CACTO.control_arr, action.reshape(1,self.conf.nb_action)])
+            CACTO.state_arr = np.vstack([CACTO.state_arr, state_next.reshape(1,self.conf.nb_state)])
+            
+            CACTO.x_ee_arr.append(self.env.get_end_effector_position(CACTO.state_arr[-1,:])[0])
+            CACTO.y_ee_arr.append(self.env.get_end_effector_position(CACTO.state_arr[-1,:])[1])
+
+            ep_arr.append(rwrd)
+
+            if step_counter==CACTO.NSTEPS_SH-1:
+                DONE = 1
+
+            # Store transition if you want to use 1-step TD
+            if self.conf.TD_N==0:
+                prioritized_buffer.add(state_prev, next_TO_action, action, rwrd, state_next, float(DONE))
```

Comment:
I'm a bit confused by the quantities you store in the buffer, and their order too. Why do you store the `next_TO_action`? Then current action should be the only one you need for CACTO. And why the next action comes before the current one in the order? I would follow the standard SARSA convention: current state, current action, reward, next state, next action.